src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java

```diff
@@ -0,0 +1,381 @@
+            }
+        }
+        return sb.toString();
+    }
+
+    private int countUnavailable(CKA[] templ) {
+        int count = 0;
+        for (CKA cka : templ) {
+            if (cka.ulValueLen == UNAVAILABLE_INFORMATION) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /**
+     * @return build result from fetch process.
+     */
+    private CKA[] buildResult() {
+        CKA[] result = new CKA[entries.length];
+        for (int i = 0; i < entries.length; i++) {
+            result[i] = entries[i].cka;
+        }
+        return result;
+    }
+
+    /**
+     * Process available attributes from the response.
+     * <p>
+     * Either an attribute is fetched or its length is now known.
+     *
+     * @param response to extract attributes from.
+     */
+    private void processAvailable(CKA[] response) {
+        for (CKA cka : response) {
+            Entry entry = getEntry(cka.type);
+            // valid attribute has length of 0, or it has a value
+            if (cka.ulValueLen == 0 || (cka.ulValueLen > 0 && cka.pValue != null)) {
+                // attribute is now fetched
+                entry.cka = cka;
+            } else if (cka.ulValueLen > 0) {
+                // pValue is null but attribute size is now known
+                entry.length = (int) cka.ulValueLen;
+            }
+        }
+    }
+
+    /**
+     * Process unavailable attributes from the response as "buffer too small".
+     * <p>
+     * This will reset sizes all attributes that are unavailable to 0, causing their lengths to be
+     * queried first before fetching.
+     *
+     * @param response to extract attributes from.
+     */
+    private void bufferTooSmall(CKA[] response) {
+        for (CKA cka : response) {
+            Entry entry = getEntry(cka.type);
+            if (cka.ulValueLen == UNAVAILABLE_INFORMATION) {
+                entry.length = 0;
+            }
+        }
+        // let's mark switch of strategy
+        useMaxLength = false;
+    }
+
+    /**
+     * Process unavailable attributes from the response as invalid.
+     * <p>
+     * This will use the resulting unavailable attributes as results (that are invalid).
+     *
+     * @param response to extract attributes from.
+     */
+    private void attributesInvalid(CKA[] response) {
+        for (CKA cka : response) {
+            Entry entry = getEntry(cka.type);
+            if (cka.ulValueLen == UNAVAILABLE_INFORMATION) {
+                entry.cka = cka;
+            }
+        }
+    }
+
+    private int maxAttributeLength(long cka) {
+        if (cka == MODULUS
+                || cka == CKA.PRIME_1
+                || cka == CKA.PRIME_2
+                || cka == CKA.EXPONENT_1
+                || cka == CKA.EXPONENT_2
+                || cka == CKA.COEFFICIENT
+                || cka == CKA.PRIVATE_EXPONENT
+                || cka == CKA.VALUE
+                || cka == CKA.EC_POINT
```

Comment:
I'd like to see CKA.DATA and CKA.DATA_BITS here. Used for certificates, PQC public keys, and some other objects.

Reply:
A Dilithium5 key (ML-DSA in the future), typically (probably when it's standardized) stored in CKA.DATA is 2592 bytes.
Dilithium2 - 1312 bytes, Dilithium 3 - 1952 bytes, Dilithium 5 - 2592 bytes

Reply:
I would gladly add them, except there is no CKA const defined for those.

Reply:
My mistake. CKA.VALUE and CKA_VALUE_BITS. Albeit I must admit I've only seen CKA.VALUE used, so ignore VALUE_BITS. I guess it depends on the use case. CKA_VALUE is used a lot for symmetric keys, key wrapping and such when it's rather small. But also for certificates and larger public keys. 
Would be a trade-off depending on use case wouldn't it?

Reply:
I really don't have anything more to contribute here. In context of a stateless GetAttributeValue call VALUE may really be anything (secret or cert). However how frequent you are accessing actual VALUE of a secret? I would say never since it is always SENSITIVE. Therefore VALUE here applies more to certificates than secret keys.